Error raised when accessing the final timestep of a dataset opened using both time range subset and interpolate_frequency

When trying to access the final timestep of an anemoi-dataset opened using a combination of `interpolate_frequency` and `start`/`end` options an `AssertionError` is raised.

**Version number**
I am using anemoi-datasets `6f6e964c14c78caa062726d670e1f3ab2b8aee0e`

**To Reproduce**
Steps to reproduce the behavior:
1. Open a gridded anemoi dataset with `start`, `end` and `interpolate_frequency` options `d = anemoi.datasets.open_dataset("aifs-ea-an-oper-0001-mars-o48-2020-2021-6h-v1.zarr", start="2020-02-01", end="2020-02-29", interpolate_frequency="1h")`
3. Access the final time-step `d[-1]` or `d[len(d)-1]`
```
File /data/users/stuart.fox/anemoi-datasets/src/anemoi/datasets/usage/gridded/interpolate.py:134, in InterpolateFrequency.__getitem__(self, n)
    130     n += self._len
    132 if n == self._len - 1:
    133     # Special case for the last element
--> 134     return self.forward[-1]
    136 i = n // self.ratio
    137 x = n % self.ratio

File /data/users/stuart.fox/anemoi-datasets/src/anemoi/datasets/usage/gridded/subset.py:181, in Subset.__getitem__(self, n)
    178 if isinstance(n, slice):
    179     return self._get_slice(n)
--> 181 assert n >= 0, n
    182 n = self.indices[n]
    183 return self.dataset[n]

AssertionError: -1
```

**Expected behavior**
The final timestep of the data should be returned.

--- a/src/anemoi/datasets/usage/gridded/indexing.py
+++ b/src/anemoi/datasets/usage/gridded/indexing.py
@@ -30,6 +30,7 @@ def _tuple_with_slices(t: TupleIndex, shape: Shape) -> tuple[TupleIndex, tuple[i
     Returns:
     Tuple[TupleIndex, Tuple[int, ...]]: A tuple containing the modified index and the changes.
     """
+    t = tuple(i + shape[j] if isinstance(i, int) and i < 0 else i for (j, i) in enumerate(t))
     result = tuple(slice(i, i + 1) if isinstance(i, int) else i for i in t)
     changes = tuple(j for (j, i) in enumerate(t) if isinstance(i, int))
     result = tuple(slice(*s.indices(shape[i])) for (i, s) in enumerate(result))

--- a/src/anemoi/datasets/usage/gridded/subset.py
+++ b/src/anemoi/datasets/usage/gridded/subset.py
@@ -178,12 +178,14 @@ class Subset(Forwards):
         if isinstance(n, slice):
             return self._get_slice(n)
 
-        assert n >= 0, n
+        if n < 0:
+            n += len(self.indices)
         n = self.indices[n]
         return self.dataset[n]
 
     def get_aux(self, n: FullIndex) -> NDArray[Any]:
-        assert n >= 0, n
+        if n < 0:
+            n += len(self.indices)
         n = self.indices[n]
         return self.dataset.get_aux(n)
 

--- a/tests/test_data.py
+++ b/tests/test_data.py
@@ -820,6 +820,43 @@ def test_subset_8() -> None:
 
 
 @mockup_open_zarr
+def test_subset_negative_indexing() -> None:
+    """Test negative indexing on a subset, including tuple indexes."""
+    ds = open_dataset("test-2021-2023-1h-o96-abcd", start=2022, end=2022)
+    assert isinstance(ds, Subset)
+    full = ds[:]
+
+    for index in (
+        -1,
+        -2,
+        -3,
+        slice(-3, None),
+        slice(-5, -2),
+        (-1,),
+        (-1, 0),
+        (-2, -1),
+        (-3, 1, 0),
+        (0, -1),
+        (slice(2, 5), -1),
+        (-1, -1, -1, -1),
+        (-1, slice(-3, None)),
+    ):
+        expected = full[index]
+        result = ds[index]
+        assert result.shape == expected.shape, index
+        assert (result == expected).all(), index
+
+
+@mockup_open_zarr
+def test_interpolate_frequency_negative_indexing() -> None:
+    """Test negative indexing on an interpolated-frequency subset."""
+    ds = open_dataset("test-2021-2023-6h-o96-abcd", start=2022, end=2022, interpolate_frequency="1h")
+    assert (ds[-1] == ds[len(ds) - 1]).all()
+    assert (ds[-2] == ds[len(ds) - 2]).all()
+    assert (ds[-3] == ds[len(ds) - 3]).all()
+
+
+@mockup_open_zarr
 def test_select_1() -> None:
     """Test selecting variables from a dataset (case 1)."""
     test = DatasetTester("test-2021-2021-6h-o96-abcd", select=["b", "d"])

